### ISSUE TITLE
languages, bracket matching, and auto-pairing

### ISSUE THREAD

**Issue by cmyr** [google/xi-editor#647]
As I've been thinking about language support, I've been thinking about two relatively minor features, bracket matching and auto-pairing with selections. By _bracket matching_ I mean the ability for the editor to identify the partner of certain characters, e.g. to highlight (or navigate to) the opening `{` that is paired with the `}` that's under the cursor; and by _auto-pairing_ I mean the ability of the editor to automatically surround a selection in a matched pair of characters when the first of those characters is entered, e.g. selecting |this| and typing a `{` might result in {this}, or a `_` might result in \_this\_.

Although these are language specific, they're both features I think could maybe live in core, and could have their behaviour modified by language-specific config overrides.

My rationale is that these both feel like fundamental features, which have a fairly simple fairly constant implementation, with the only variables being the exact characters that get this treatment; for instance in HTML we would want to be able to match the `<` character to the `>` character, which we would not want to do in a programming language where `<` was only ever an operator.

**Comment by vlovich** [comment]
Yeah, that seems like a core algorithm to me language-specific plugins can
opt-in to use. Each language would just be able to provide the candidate
set of pairs of characters (glyphs? Not sure of the Unicode terminology)
for matches.  Is that how you were thinking about it?

It’s probably a good idea to let plugins be able to not opt-in and do their
own thing as there are “languages” that aren’t as regular and may have
different logic (markdown?)

**Comment by vlovich** [comment]
Yup. That was my thinking

**Comment by lord** [comment]
I might try to take a stab at this!

**Comment by lord** [comment]
A couple open design questions I'm thinking about:

Are there use cases for the list of autopairing characters to be different from bracket matching? In TextMate, [they are the same](https://macromates.com/manual/en/preferences_items#paired-characters), and in some aspects I think of them as the same feature, but also maybe there are reasons to keep them separate that I'm not thinking of.

Beyond just the opening character, there's a couple other bits of information. The closing character is one. But there's also a bit that determines if whitespace is required before the insertion point. For instance, in Sublime:
`hello|` -> type " -> `hello"|`
but
`hello|` -> type { -> `hello{|}`
I guess the question I'm thinking about here is if languages would want to customize these aspects of each character, or if it's acceptable to just hardcode them in on a per-character level, and then allow languages to just turn characters on and off.

**Comment by cmyr** [comment]
VSCode does something similar to this, so I might take a look [at that](https://code.visualstudio.com/docs/extensionAPI/extension-points#_contributeslanguages).

For the second bit: I think these are separate features.  What I'm thinking of right now is what VSCode calls `surroundingPairs`; that is, characters that will surround a selection when the leading character is typed.

For `autoclosingPairs`, I'm still slightly undecided about who's job this is, so I want to sit on it for a few more days. I'm going to be hacking on syntect-based auto-indent next, and that will freshen my memory and generally (hopefully) clarify my thinking. The basic blocker is that syntect and VSCode handle this in slightly different and irreconcilable ways.

So I would focus for now on the narrower scope of 'surroundingPairs', and I'll try and come up with a more lucid explanation of how I'm thinking about auto-closing.

**Comment by lord** [comment]
Ok, I'm not sure this is possible without edit priority? If I submit a normal edit that adds the two characters, the selection looks like this (between the bars):

    {|texthere}|
     ^         ^ selection

If we somehow were to disable `after`, we could get this output:

    |{texthere|}
    ^          ^ selection

We could also do `keep_selection`, which would give us:

    |{texthere}|
    ^          ^ selection

I think there's a pretty tricky case here, though. Imagine there are *two* selections:

    |text|here|
    ^    ^       selection A
         ^    ^  selection B

Selection A ends at the same point that selection B starts. In Sublime, if I type `{`, I get:

    {|text|}{|here|}
    ^     ^         selection A
             ^    ^ selection B

I don't fully understand the plans for edit priority, so maybe this is solved by it? But in the case above, I'd want my edit to go before some selection boundaries, and after some other selection boundaries, and also ensure the `}{` appears in the right order. In my head, not sure if this is helpful, but I think of this kind of edit less in terms of 'before' and 'after' and instead in terms of 'inside' and 'outside' the selection.

**Comment by cmyr** [comment]
@lord this is interesting, and probably worth pinging @raphlinus over. I wasn't sure that you _could_ have a selection `|one|two|`; I thought the two regions might be merged. Seems not, although they _do_ get merged on transpose, which we'd have to call a bug.

It might be worth reading through #933 and see if that seems relevant; I haven't been following that issue too closely.

**Comment by lord** [comment]
Yea, I thought they might get merged too, but I can definitely see the use case for adjacent selections?

In any case, I think I found a possible solution by extending `keep_selections`? See #968

**Comment by raphlinus** [comment]
I have a bunch of responses, ranging from high level to granular.

I _don't_ think that bracket matching is a simple case of matching bracket characters. You have to do at least some level of syntax analysis of the input, because characters inside strings and comments don't match against those outside. My original high level conception is that you bundle basically _all_ operations that depend on syntax analysis in a "language" plugin, which bundles a number of tasks I would consider related: highlighting, computation of indent, bracket matching, and (this last is less clear) code folding. But maybe that's not a good architecture. Another path to go forward is to have syntax highlighting scopes explicit in the data model, and use scopes to resolve many of these questions. But that approach has its own problems, including races because the scopes are updated async; if you type too fast, then the logic for deciding how to respond to a character might be looking at stale scope info.

I do think adjacent selections are useful. Sublime definitely has the behavior, and I adapted it from that. The fact that transpose merges them is a bug.

I think priority can address the issues raised in the `texthere` comment. One of the things that makes it interesting is that you have to have a compound edit with multiple insertions, each with its own priority. But that's the same situation as whitespace insertion. I haven't actually wrapped my mind around this deeply enough to figure out whether priority can handle all the cases (including transpose, which we now know is buggy), or whether there's scope for expanding keep_selections into "drift" as proposed in #968.

Posting these comments now because I think more discussion is useful, even though there's obviously more to think through.

**Comment by raphlinus** [comment]
I discussed this with @cmyr and @scholtzan. The example of trying to insert between two adjacent selections is a good one; this is not a case I'd considered, and it's not at all clear to me how priority would deal with it. The approach in #968 does seem like a pragmatic approach, and I'm ok going forward with that. I think we'll learn something from it, and if we do come up with a more principled solution, we can migrate to that later.

@cmyr also pointed out that, for the specific case of inserting paired characters, we can afford to have imprecise syntax information. I agree, but would like to have a more general solution, for the other cases mentioned in the original issue.

**Comment by lord** [comment]
Definitely agree that auto pairing should have much better syntax info, but yea, just working on surrounding pairs for now. And yea, the adjacent selections thing is weird!

Since it sounds like InputDrift is something we'll try out for now, I'll clean up #968 with some tests and better comments, and ping that PR when it's ready for review.

**Comment by jansol** [comment]
Should this be closed in favor of #933?
